keps/sig-node/2238-liveness-probe-grace-period/README.md

```diff
@@ -0,0 +1,491 @@
+decouple.
+
+[Other implementation options](#alternatives) might avoid an API change, but
+would end up breaking backwards compatibility which some users may have come to
+[expect/rely on](https://github.com/kubernetes/kubernetes/issues/64715#issuecomment-756100727).
+
+An example of a scenario where this bug was apparent is a long outage reported
+by an ingress controller:
+
+> the controller uses a 3600s grace period so that it can drain long requests
+> from end user app traffic (it listens on endpoints that can remain available
+> for hours like host network, but also incoming tcp load balancers which have
+> open connections), but when the controller wedges it takes an hour to
+> resolve, which is the worst possible outcome (the process wedges immediately,
+> but kube has to wait the full hour to kill it, defeating the goal of
+> liveness). [comment](https://github.com/kubernetes/kubernetes/issues/64715#issuecomment-756201502)
+
+### Goals
+
+- Liveness probes can have a configurable timeout on failure.
+- Existing behaviour is not changed.
+
+### Non-Goals
+
+Any changes to probe behaviour outside of the scope of addressing the
+[bug](https://github.com/kubernetes/kubernetes/issues/64715) this is intended
+to address.
+
+## Proposal
+
+We can add a `gracePeriodSeconds` integer field to the Probe API. This field
+would specify how long the kubelet should wait before forcibly terminating the
+container when a probe fails.
+
+This change would be [API-compatible][compatible]. If the field is not
+specified, we will default to the current behaviour, which uses the
+`terminationGracePeriodSeconds`.
+
+[compatible]: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility
+
+### Configuration example
+
+```yaml
+spec:
+  terminationGracePeriodSeconds: 3600
+  containers:
+  - name: test
+    image: ...
+
+    ports:
+    - name: liveness-port
+      containerPort: 8080
+      hostPort: 8080
+
+    livenessProbe:
+      httpGet:
+        path: /healthz
+        port: liveness-port
+      failureThreshold: 1
+      periodSeconds: 60
+      # New field #
+      gracePeriodSeconds: 60
+```
+
+### User Stories (Optional)
+
+N/A - bugfix
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+This should be a low-risk API change as it is backwards-compatible. If the
+field is not set, the current behaviour will be maintained.
+
+## Design Details
+
+### Test Plan
+
+This change will be unit tested for the API changes and
+backwards-compatibility.
+
+If added to the Probe API, this will also need to be tested in the readiness
+and startup probe cases.
+
+E2E integration tests will verify the correct behaviour for the current broken
+use case, where a `terminationGracePeriodSeconds` is set to a
+large value, a probe fails, and with `livenessProbe.gracePeriodSeconds` set,
+the container terminates quickly.
```

Comment:
do we allow livenessProbe.gracePeriodSeconds to be higher than terminationGracePeriodSeconds? do we need some kinds of validation?